lib/cn/kvset.c

```diff
@@ -3633,6 +3674,112 @@ kvset_minkey(struct kvset *ks, const void **minkey, u16 *minklen)
     *minklen = ks->ks_minklen;
 }
 
+struct kvset_vgroup_map *
+kvset_vgmap_alloc(uint32_t nvgroups)
+{
+    struct kvset_vgroup_map *vgmap;
+    size_t sz = sizeof(*vgmap) + 3 * nvgroups * sizeof(*(vgmap->vbidx_out));
+
+    vgmap = calloc(1, sz);
+    if (vgmap) {
+        vgmap->nvgroups = nvgroups;
+        if (nvgroups > 0) {
+            vgmap->vbidx_out = (void *)(vgmap + 1);
+            vgmap->vbidx_adj = vgmap->vbidx_out + nvgroups;
+            vgmap->vbidx_src = vgmap->vbidx_adj + nvgroups;
+        }
+    }
+
+    return vgmap;
+}
+
+void
+kvset_vgmap_free(struct kvset_vgroup_map *vgmap)
+{
+    free(vgmap);
+}
+
+/* TODO: Use binary search as an optimization */
+int
+kvset_vgmap_vbidx_out2adj(struct kvset_vgroup_map *vgmap, int vbidx_out)
+{
+    int i;
```

Comment:
The input type `int` matches the return type `int`, which makes sense.  The return type is `int` so it can be overloaded and return `-1` on errors.   But callers don't even check for errors.  I would avoid overloading the return value and return merr_t on failure, return result in output param, make types match.

And handle the error.  Or use abort instead of assert and have the funcition never return error.

The safest thing is handle the error and fail the compaction but let the system continue running (assuming this is on a compaction operation's call stack).

Reply:
I get that, but the type in the omf is a uint32_t, so an int doesn't have the same bounds as uint32_t.

Reply:
the end result of my comment was i agree with you :)

Reply:
These functions now return an merr_t. The callers check for error and either pass it up the stack or abort.